Add a 'noopener' link attribute

mikewest opened the pull request:
Based on Charlie Reis' proposal at [1], this patch adds a mechanism
by which web developers can open a new window without granting that
window access to the opener browsing context. This is possible today
via , but that is both obscure and
poorly explanatory. The  attribute clarifies the behavior.

[1]: https://wiki.whatwg.org/wiki/Links_to_Unrelated_Browsing_Contexts

mikewest commented:
Getting something like this into HTML might make it possible for Secure Contexts to distinguish between popups that could be secure contexts and popups that couldn't. WDYT, @annevk?

annevk commented:
I think we should add some text that explains why you want to use this instead of `noreferrer`. Or the other way around.

mikewest commented:
Simplified the patch, and added some text to `noreferrer` to note that it implies `noopener` in specific circumstances. Is that what you're looking for?

Basically, we just tacked completely unrelated functionality onto `noreferrer` at some point in ~2011. This is an attempt to explain the behavior, and make it available separately.

annevk commented:
"Charlie Reis" needs to be added to the acknowledgements section too.

domenic commented:
Can we also make sure a caniuse entry and web platform tests are in progress?

Do we have interest from more than one implementer?

mikewest commented:
@rlbmoz, @dveditz: How do y'all feel about implementing this in Firefox? :)

mikewest commented:
Filed https://www.chromestatus.com/feature/5651874132787200, FWIW.

annevk commented:
We need to credit @sc0ttbeardsley in Acknowledgements since he proposed this in https://lists.w3.org/Archives/Public/public-whatwg-archive/2015May/0046.html. Per http://logs.glob.uno/?c=mozilla%23content&s=5+Nov+2015&e=5+Nov+2015#c338410 at least some from Mozilla are on board with this feature.

annevk commented on source:
```diff
@@ -21971,14 +21978,30 @@ interface <dfn>HTMLHyperlinkElementUtils</dfn> {
-  <p>This keyword also <a href="#noopener">causes the <code data-x="dom-opener">opener</code>
-  attribute to remain null</a> if the hyperlink <span data-x="creating a new browsing context">creates</span> a new <span>browsing context</span>.</p>
+  <p class="note">For historical reasons, the <code data-x="rel-noreferrer">noreferrer</code>
+  attribute implies the behavior associated with the <code data-x="rel-noopener">noopener</code>
```
These are not the attributes you're looking for. (Keywords.)

mikewest replied:
Thanks, fixed.

mikewest commented:
I added the `window.opener` changes I'd like to make, since they touched basically the same bits of the spec. Happy to split it out if you decide that it's simpler that way.

sc0ttbeardsley commented:
Thanks for the tag @annevk. I have been chatting with Steve Workman at Mozilla (sorry, don't know his github) and his team was going to look into at least throwing a warning when window.opener is modified across origins (I think IE does this already).

@mikewest Here is the original bug I filed: https://www.w3.org/Bugs/Public/show_bug.cgi?id=28821 (I suspect this PR should resolve that). Also it looks like http://crbug.com/69267 is related to http://crbug.com/168988

I just closed my PR on the wrong repo.

Regarding the proposal, I had thought about using "noopener" keyword but I think "newcontext" might be a bit more explicit. Ultimately we don't *only* want window.opener to be nullified, right? I think there is a desire to also do all of the other things (it might be hard to explicitly list everything though) associated with a new browsing context.

annevk commented:
Note to self: file a bug against Gecko once this is merged.

annevk commented on source:
```diff
@@ -21980,14 +21987,30 @@ interface <dfn>HTMLHyperlinkElementUtils</dfn> {
+  <p class="note">For historical reasons, the <code data-x="rel-noreferrer">noreferrer</code>
+  keyword implies the behavior associated with the <code data-x="rel-noopener">noopener</code>
+  keyword when present on a hyperlink that <span
+  data-x="creating a new browsing context">creates</span> a new <span>browsing context</span>.
```
You are allowed to break inside an attribute value when it has spaces.

mikewest replied:
This document has seriously strange style rules.

annevk commented on source:
```diff
@@ -78389,33 +78412,43 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+       new start for session storage</dfn>.</p>
+
+       <p class="note">If it is immediately <span data-x="navigate">navigated</span>, then the
+       navigation will be done with <span>replacement enabled</span>.</p></dd>
```
Deduplicate this note by placing it below the `</dl>` and precede it with "Either way, if it is..."?

mikewest replied:
Refactoring this completely.

annevk commented on source:
```diff
@@ -78389,33 +78412,43 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
-
-     <p class="note">If it is immediately <span data-x="navigate">navigated</span>, then the
-     navigation will be done with <span>replacement enabled</span>.</p></dd>
+     create a new browsing context:</dt>
```
I wonder if the requirements here would be better fixed by simply disowning the opener if one of the keywords is present and then fixing #313 so that disowning the opener does the correct thing. Or did you have a different fix for that issue in mind?

mikewest replied:
I think that's a reasonable approach. I think it's also a good idea to move the work out of this algorithm, and into the `window.open` and "following hyperlinks" algorithms. That should make it a bit clearer what's actually going on.

annevk commented:
@sc0ttbeardsley what other things are you thinking of that are not related to the browsing context having an opener browsing context (and therefore being auxiliary)?

sc0ttbeardsley commented:
@annevk <a href="http://crbug.com/69267">the original chromium bug</a> mentioned running in a new "process". We should avoid getting so explicit that we dive into browser implementation variations on what constitutes a new browsing context. For instance, do all browsers equate new processes with new browsing contexts?

annevk commented:
Either way it's a new browsing context. When the new browsing context has no opener it can run (in theory) in a distinct process since there's no synchronous script access to it.

mikewest commented:
Refactored the patch a bit. WDYT, @annevk?

@sc0ttbeardsley: Thanks for the links to the chromium bugs. I've added some labels there.

annevk commented on source:
```diff
@@ -21468,6 +21473,13 @@ interface <dfn>HTMLHyperlinkElementUtils</dfn> {
+     <td><code data-x="rel-noopener">noopener</code></td>
+     <td><em>not allowed</em></td>
+     <td><span data-x="hyperlink annotation">Annotation</span></td>
+     <td>Requires that any <span>browsing context</span> created by following the hyperlink must not have a <span>opener browsing context</span>.</td>
```
an*

mikewest replied:
Indeed!

annevk commented on source:
```diff
@@ -20968,6 +20968,11 @@ interface <dfn>HTMLHyperlinkElementUtils</dfn> {
 
    </li>
 
+   <li><p>If <var>subject</var>'s <a href="#linkTypes">link types</a> include the <code
```
includes*

mikewest replied:
Hrm. No? "types" is plural, so it doesn't "include". At least, "types includes" sounds strange.

annevk replied:
Agreed. Misread.

annevk commented:
I think with those nits this looks good. After this lands #313 will still need to be fixed to fully define the behavior added here accurately, correct?

mikewest commented:
Squished. I didn't change "includes", but I will if you feel strongly about it.

annevk commented:
Ah, unfortunately this was not rebased so it did not merge completely cleanly. https://github.com/whatwg/html/commit/2992ea921bc75e44157451a37a807a8ce0b9a884 is the commit.

annevk commented:
Thank you though for your efforts!

annevk commented:
Gecko bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1222516.

sc0ttbeardsley commented:
@annevk sorry, I see this is "closed" was the change *not* merged for some reason?

awebdeveloper commented:
webkit bug: https://bugs.webkit.org/show_bug.cgi?id=155166

awebdeveloper commented:
IE bug: https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/12942405-implement-rel-noopener
